Sources/SwiftProtobuf/Message+FieldMask.swift

```diff
@@ -0,0 +1,136 @@
+// Sources/SwiftProtobuf/Message+FieldMask.swift - Message field mask extensions
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extend the Message types with FieldMask utilities.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+extension Message {
+
+  /// Checks whether the given path is valid for Message type.
+  ///
+  /// - Parameter path: Path to be checked
+  /// - Returns: Boolean determines path is valid.
+  public static func isPathValid(
+    _ path: String
+  ) -> Bool {
+    var message = Self()
+    return message.hasPath(path: path)
+  }
+
+  internal mutating func hasPath(path: String) -> Bool {
+    do {
+      try set(path: path, value: nil, mergeOption: .init())
+      return true
+    } catch let error as PathDecodingError {
+      return error != .pathNotFound
+    } catch {
+      return false
+    }
+  }
+
+  internal mutating func isPathValid(
+    _ path: String
+  ) -> Bool {
+    hasPath(path: path)
+  }
+}
+
+extension Google_Protobuf_FieldMask {
+
+  /// Defines available options for merging two messages.
+  public struct MergeOptions {
+
+    public init() {}
+
+    /// The default merging behavior will append entries from the source
+    /// repeated field to the destination repeated field. If you only want
+    /// to keep the entries from the source repeated field, set this flag
+    /// to true.
+    public var replaceRepeatedFields = false
+  }
+}
+
+extension Message {
+
+  /// Merges fields specified in a FieldMask into another message.
+  ///
+  /// - Parameters:
+  ///   - source: Message that should be merged to the original one.
+  ///   - fieldMask: FieldMask specifies which fields should be merged.
+  public mutating func merge(
+    with source: Self,
+    fieldMask: Google_Protobuf_FieldMask,
+    mergeOption: Google_Protobuf_FieldMask.MergeOptions = .init()
+  ) throws {
+    var visitor = PathVisitor<Self>()
+    try source.traverse(visitor: &visitor)
+    let values = visitor.values
+    for path in fieldMask.paths {
```

Comment:
Maybe put a note/todo here, since this seems like this could be a performance issue in the future; if there is a way to traverse once setting the any requested fields that seems like it would be better.

Reply:
It already gets all values of a message with only one traverse. But we can improve it with only getting required paths instead of all paths.